src/sentry/rules/actions/base.py

```diff
@@ -83,56 +90,152 @@ def get_form_instance(self):
         return self.form_cls(self.data, integrations=self.get_integrations())
 
 
+def _linked_issues(event, integration):
+    return ExternalIssue.objects.filter(
+        id__in=GroupLink.objects.filter(
+            project_id=event.group.project.id,
+            group_id=event.group.id,
+            linked_type=GroupLink.LinkedType.issue,
+        ).values_list("linked_id", flat=True),
+        integration_id=integration.id,
+    )
+
+
+def get_linked_issue_ids(event, integration):
+    return _linked_issues(event, integration).values_list("key", flat=True)
+
+
+def has_linked_issue(event, integration):
+    return _linked_issues(event, integration).exists()
+
+
+def create_link(key, integration, installation, event):
+    external_issue = ExternalIssue.objects.create(
+        organization_id=event.group.project.organization.id,
+        integration_id=integration.id,
+        key=key,
+        title=event.title,
+        description=installation.get_group_description(event.group, event),
+    )
+    GroupLink.objects.create(
+        group_id=event.group.id,
+        project_id=event.group.project.id,
+        linked_type=GroupLink.LinkedType.issue,
+        linked_id=external_issue.id,
+        relationship=GroupLink.Relationship.references,
+        data={"provider": integration.provider},
+    )
+
+
+def build_description(event, rule_id, installation, generate_footer):
+    """
+    Format the description of the ticket/work item
+    """
+    project = event.group.project
+    rule_url = u"/organizations/{}/alerts/rules/{}/{}/".format(
+        project.organization.slug, project.slug, rule_id
+    )
+
+    return installation.get_group_description(event.group, event) + generate_footer(rule_url)
+
+
+def create_issue(event, futures):
+    """Create an issue for a given event"""
+    organization = event.group.project.organization
+
+    for future in futures:
+        rule_id = future.rule.id
+        data = future.kwargs.get("data")
+        provider = future.kwargs.get("provider")
+        integration_id = future.kwargs.get("integration_id")
+        generate_footer = future.kwargs.get("generate_footer")
+
+        try:
+            integration = Integration.objects.get(
+                id=integration_id,
+                provider=provider,
+                organizations=organization,
+                status=ObjectStatus.VISIBLE,
+            )
+        except Integration.DoesNotExist:
+            # Integration removed, rule still active.
+            return
+
+        installation = integration.get_installation(organization.id)
+        data["title"] = event.title
+        data["description"] = build_description(event, rule_id, installation, generate_footer)
+
+        # HACK to get fixVersion in the correct format
+        if data.get("fixVersions"):
```

Comment:
@ceorourke Is this a Jira specific hack? I fear this is going to break with other form fields such as a component that are lists. Either way, I would encapsulate this behavior into a method like `fix_data_for_issue` that can be implemented differently for each integration.